fms_extras/models/paged_gpt_bigcode.py

```diff
@@ -0,0 +1,383 @@
+        self.position_embedding = nn.Embedding(self.config.max_pos, self.config.emb_dim)
+
+        self.dec_norm = nn.LayerNorm(self.config.emb_dim, eps=self.config.ln_eps)
+
+        if self.config.emb_dropout:
+            self.emb_dropout = nn.Dropout(self.config.emb_dropout)
+
+        if self.config.p_dropout:
+            self.dropout = nn.Dropout(self.config.p_dropout)
+
+    def forward(
+        self,
+        x: torch.LongTensor,
+        mask: Optional[torch.Tensor] = None,
+        cache_data: Optional[PagedAttentionCacheData] = None,
+        use_cache: bool = False,
+        attn_algorithm: Optional[str] = None,
+    ):
+        # Embed the given vocabulary indices using the given attention mask, with pre-/post-norm and dropout as specified
+        # x_in: batch_size x seq_len
+        # mask: batch_size x seq_len x seq_len
+        # bias: nheads x seq_len x seq_len
+
+        qlen = x.size(1)
+        filled_cache = False
+
+        # if we are using the cache, the key length needs to be extended with the past keys length
+        if use_cache:
+            if cache_data:
+                filled_cache = cache_data.is_filled()
+
+        # if mask is none, we need to specify causal mask
+        if mask is None:
+            # we are caching and can assume all 1s in the mask
+            if use_cache and filled_cache and qlen == 1:
+                # b x h x qlen x kvlen
+                is_causal_mask = False
+            else:
+                is_causal_mask = True
+        else:
+            is_causal_mask = False
+
+        x_emb = self.embedding(x)
+
+        # if pad_id exists
+        #   is_pad will be a BoolTensor
+        #   otherwise pad_id will not be taken into account
+        if self.config.pad_id is None:
+            is_pad = torch.zeros_like(x, dtype=bool, device=x.device)
+        else:
+            is_pad = x == self.config.pad_id
+
+        if cache_data is None or cache_data.position_ids is None:
+            position_ids = ((~is_pad).cumsum(1) - 1).clamp(min=0)
+
+            if cache_data is not None:
+                cache_data.position_ids = position_ids
+        else:
+            position_ids = cache_data.position_ids
+
+        # look up position embeddings
+        position_out = self.position_embedding(position_ids)
+
+        # zero out the associated position embeddings
+        if self.config.pad_id is not None:
+            position_out = position_out.mul(~is_pad.unsqueeze(-1))
+
+        # perform absolute position embedding
+        x = x_emb + position_out
+
+        # apply dropout to embeddings
+        if self.config.emb_dropout:
+            x = self.emb_dropout(x)
+
+        # this is the output cache for all the decoder layers
+        present_key_value_states = []
+
+        for i, layer in enumerate(self.layers):
+            output = layer(
+                x=x,
+                mask=mask,
+                cache_data_layer=None
```

Comment:
the value for the `line` should be assigned to a variable first or something, otherwise this looks like it says `cache_data_layer=None`